src/recode.jl

```diff
@@ -52,22 +54,47 @@ A user defined type could override this method to define an appropriate test fun
 optimize_pair(pair::Pair) = pair
 optimize_pair(pair::Pair{<:AbstractArray}) = Set(pair.first) => pair.second
 
+function missing_check(value)
+    ismissing(value) && throw(MissingException("missing value found, but dest does not support them: " *
+                                       "recode them to a supported value"))
+    value
+end
+
+function recode!(dest::AbstractArray{T}, src::CategoricalArray, default::Any, pairs::Pair...) where {T}
+    if length(dest) != length(src)
+        throw(DimensionMismatch("dest and src must be of the same length (got $(length(dest)) and $(length(src)))"))
+    end
+    
+    pairs = map(pairs) do p
+        p.first => convert(T, p.second)
+    end
+    recoded = recode(src, default, pairs...)
+    if T >: Missing
+        dest .= unwrap.(recoded)
+    else
+        dest .= missing_check.(unwrap.(recoded))
+    end
```

Comment:
Rather than doing this, to avoid making a copy and two passes over the data, we should call `recode` on `levels(src)`, and then do something like:
```julia
@inbounds for i in eachindex(dest, src)
    dest[i] = newlevels[src.refs[i]+1]
end
```

The actual implementation needs to be a bit more complex so that the first entry in `newlevels` is `missing` (to handle the case when `src.refs` is 0).

Reply:
> Rather than doing this, to avoid making a copy and two passes over the data

By `copy` do you mean copy of the `src.levels`? In this implementation no copy of the actual array (or `refs`) is made which is the main reason why it is so much faster (as outlined in my [StackOverflow answer](https://stackoverflow.com/a/67571105/15281224)) is because all the actual copying of the `refs` happens only once at the last line `dest .= unwrap.(recoded)` the `recoded` variable shares the refs with `src`.

Reply:
`recode(src, default, pairs...)` allocates a new vector, right? That's relatively fast, but it's even better to avoid it.

Reply:
You're right, I remembered the details wrong. In the StackOverflow example I did:

```
mapping = Dict("X"=>1, "Y"=>2, "Z"=>3)
b = CategoricalArray{Int64,1,UInt32}(undef, 0)
b.refs = a.refs
levels!(b.pool, [mapping[l] for l in levels(a.pool)])
```
which is similar to what you're suggesting. However, in this PR we initialize the CategoricalArray that will be put in the `recoded` variable with something like `CategoricalArray{S, N, R}(undef, size(a))` so the `refs` are not shared between src and recoded.